internal/lib/command.ml

```diff
@@ -16,36 +16,89 @@
 
 (** Utilities for running commands. *)
 
+module Fun = Base.Fun
+module Option = Base.Option
+
 exception Error of string
 
 let command bin args =
   match args with
   | [] -> (Filename.quote bin)
   | _ -> Printf.sprintf "%s %s" (Filename.quote bin) (String.concat " " (List.map Filename.quote args))
 
-let run bin args =
-  let cmd = command bin args in
-  match Sys.command cmd with
-  | 0 -> ()
-  | n -> raise (Error (Printf.sprintf "Process returned error code %i" n))
-
-let run_with_stdout bin args f =
-  let cmd = command bin args in
-  let stdout = Unix.open_process_in cmd in
-  let ret = f stdout in
-  match Unix.close_process_in stdout with
-  | Unix.WEXITED 0 -> ret
-  | Unix.WEXITED n -> raise (Error (Printf.sprintf "Process returned error code %i" n))
-  | Unix.WSIGNALED n -> raise (Error (Printf.sprintf "Process was killed by signal %i" n))
-  | Unix.WSTOPPED n -> raise (Error (Printf.sprintf "Process was stopped by signal %i" n))
 
-let run_with_stdin_and_stdout bin args in_f out_f =
-  let cmd = command bin args in
-  let stdout, stdin = Unix.open_process cmd in
-  in_f stdin ;
-  let ret = out_f stdout in
-  match Unix.close_process (stdout, stdin) with
-  | Unix.WEXITED 0 -> ret
+(* Sentinel values for the optional function parameters to [run]. *)
+let unset_in (_ : out_channel) = failwith "unset_in"
+let unset_out (_ : in_channel) = failwith "unset_out"
+let unset_err (_ : in_channel) = failwith "unset_err"
+
+let run ?stdin:(in_f=unset_in) ?stdout:(out_f=unset_out) ?stderr:(err_f=unset_err) bin args =
```

Comment:
Hi Ethel,

Are you aware of another way to use optional arguments?
```OCaml
let run ?stdin:in_f ... =
    ...
    let in_fd,in_f,close_in_pipe = match in_f with
    | None -> Unix.stdin,nop,nop
    | Some in_f -> 
         let fd, o = out_pipe () in
         fd, (fun _ -> Unix.close fd ; in_f o), (fun _ -> close_out o) in
```
I find this solution more elegant, since it avoids the never called sentinel values and the `==` comparison. If interested, have a look [at section 4.1.1 in OCaml manual](https://caml.inria.fr/pub/docs/manual-ocaml/lablexamples.html#ss:optional-arguments).

Reply:
That is much better! Thank you! :D

Reply:
Not that much. Thanks, merging imminent.